src/libostree/ostree-repo-libarchive.c

```diff
@@ -1086,9 +1094,26 @@ write_directory_to_libarchive_recurse (OstreeRepo *self, OstreeRepoExportArchive
             g_autoptr (GInputStream) file_in = NULL;
             g_autoptr (GFileInfo) regular_file_info = NULL;
             const char *checksum;
+            GFile *old_path;
 
             checksum = ostree_repo_file_get_checksum ((OstreeRepoFile *)path);
 
+            old_path = g_hash_table_lookup (seen_checksums, checksum);
+            if (old_path)
+              {
+                g_autofree char *old_pathstr = file_to_pathstr (root, opts, old_path);
+
+                archive_entry_set_hardlink (entry, old_pathstr);
+                if (!write_header_free_entry (a, &entry, error))
+                  goto out;
+
+                break;
+              }
+            else
+              {
+                g_hash_table_insert (seen_checksums, (char *)checksum, g_object_ref (path));
```

Comment:
This may be worth a comment like:

`/* The checksum string is owned by the repo file object */` or so; I paused for a few seconds here thinking about the memory management.